MSBuild: add argument to build() to opt-in autoconsumption of props files generated by MSBuildToolchain & MSBuildDeps

SpaceIm opened the pull request:
related to https://github.com/conan-io/conan/issues/12155#issuecomment-1252832378

Changelog: (Fix): In `MSBuild`, add an argument to `build()` method to opt-in automatic consumption of props files generated by `MSBuildToolchain` & `MSBuildDeps`, without requiring intrusive modification of vcxproj files.
Docs: https://github.com/conan-io/docs/pull/2885

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

SpaceIm committed:
MSBuild: ensure to consume props files generated by MSBuildToolchain & MSBuildDeps with highest precedence

SpaceIm commented:
I've tested with libsodium recipe (without this ugly workaround https://github.com/conan-io/conan-center-index/blob/4e0dc05121ed9999dcd3ed8841844757f6495ce0/recipes/libsodium/all/conanfile.py#L137-L171 -mandatory without the fix in this PR-, and a fix in custom `configuration` of its `MSBuildToolchain`) and it works like a charm, now vc runtime can be honored (as well as multiprocessing build).

SpaceIm committed:
rely on predictable props names from generators

SpaceIm commented:
see these conan-center PR for current workarounds:
https://github.com/conan-io/conan-center-index/pull/15041
https://github.com/conan-io/conan-center-index/pull/15043
https://github.com/conan-io/conan-center-index/pull/15046
https://github.com/conan-io/conan-center-index/pull/15050

It's more or less always the same boilerplate (boilerplate we could remove with this PR):
https://github.com/conan-io/conan-center-index/blob/4a2cbbba7fce7da91a656890744f3463c7429846/recipes/libsodium/all/conanfile.py#L130-L145
https://github.com/conan-io/conan-center-index/blob/22d52120bb27d81d551ada9255e012755b023276/recipes/xz_utils/all/conanfile.py#L104-L128
https://github.com/conan-io/conan-center-index/blob/110eb267a9776031a5189abd06254b07c88abf25/recipes/libmicrohttpd/all/conanfile.py#L137-L151
https://github.com/conan-io/conan-center-index/blob/a4ba448dc6c474198d669242ea5e1e626349581e/recipes/zimg/all/conanfile.py#L93-L108

SpaceIm committed:
adapt template & tests

SpaceIm committed:
force PlatformToolset from msbuild command line

SpaceIm committed:
move /p options first

memsharded commented on conan/tools/microsoft/msbuild.py:
```diff
@@ -27,10 +32,20 @@ def __init__(self, conanfile):
+            props_path = os.path.join(self._conanfile.generators_folder, props_file)
+            if os.path.exists(props_path):
+                props_paths.append(props_path)
+        if props_paths:
```
This cannot be made default, because that will be forcing the injection of whole dependencies into the current project, and there are many users that don't want that, but manual addition of individual dependencies into different subprojects.

At least, it should have an opt-out to allow users defining it.

SpaceIm replied:
So keep this logic for `conantoolchain.props`, but allow to not autoconsume `conandeps.props`?

Regarding what users want, I'm wondering: users who manually edit their MSBuild project files, are they really using MSBuild conan helper? Or just MSBuildToolchain & MSBuildDeps?
This PR is really about providing the same high level logic in this toolchain/helper family than all other families in conan v2. It's the only family where the build helper does not automatically consume its toolchains. There is no opt-in or opt-out in Autotools helper to not globally inject AutotoolsDeps, so why a different behavior in MSBuild helper?

memsharded replied:
From the feedback from users using this integration, it is very common that they have different subprojects and they want to inject different dependencies to different subprojects. If MSBuild pass ``conandeps.props`` by default, this cannot be done.

There are definitely other build systems that do the same, CMake uses the same pattern, users need to define ``find_package()`` and ``target_link_libraries()`` to the things that they want to use, it is not automagically applied to the whole project. For sure that is a built-in construct, but this level of control is what can be achieved with the MSBuild integration. The fact that it is not possiblel to do it in Autotools, for example, shouldn't be used as the design goal for it.

SpaceIm replied:
I mostly agree for `*Deps` generators (even thought it's transparent in case of CMake), but `MSBuild` is the only build helper not consuming automatically & transparently its "glue toolchain".

memsharded commented on conan/tools/microsoft/msbuild.py:
```diff
@@ -27,10 +32,20 @@ def __init__(self, conanfile):
-        cmd = ('msbuild "%s" /p:Configuration=%s /p:Platform=%s'
-               % (sln, self.build_type, self.platform))
+        cmd = (f'msbuild "{sln}" /p:Configuration={self.build_type} '
+               f"/p:Platform={self.platform} /p:PlatformToolset={self.toolset}")
```
This adds back the PlatformToolset as argument only, and that doesn't work for user consumption, that will only be loading the toolchain. That configuration needs to be in the toolchain file.

SpaceIm replied:
I've not removed PlatformToolset from toolchain (actually I've not modified content of files generated by toolchains). It's added as an argument to msbuild because when `conantoolchain.props` is injected with `/p:ForceImportBeforeCppTargets`, I've seen that PlatformToolset from this toolchain was ignored.

SpaceIm replied:
Actually, from my experimentation when `conantoolchain.props` is injected with `/p:ForceImportBeforeCppTargets`:
- Everything under `ItemDefinitionGroup` is honored.
- Everything under `PropertyGroup Label="Configuration"` (`PlatformToolset` & custom properties) is ignored by MSBuild.

This is why I inject `/p:PlatformToolset` in msbuild command line (and if I could, I would provide a properties argument to `msbuild` build helper to enforce properties in command line).

SpaceIm replied:
https://learn.microsoft.com/en-us/cpp/build/project-property-inheritance

> When you set a property with property pages, it's also important to pay attention to where you set it. If you set a property to "X" in a .props file, but the property is set to "Y" in the project file, then the project will build with the property set to "Y".

https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-properties

> MSBuild lets you set properties on the command line by using the -property (or -p) switch. These global property values override property values that are set in the project file. This includes environment properties, but does not include reserved properties, which cannot be changed.

memsharded replied:
The problem with ``/p:ForceImportBeforeCppTargets`` is the developer experience. Most developers with VS will just double click the project or solution Whatever is being passed should be in the ``vcxproj`` by default, it cannot rely on cli args to work. This is probably the most massive use case of this integration, there are companies out there using it, and hundreds of developers do this daily. So this cannot be changed if this doesn't keep working as-is (this is why I suggest an opt-in for ConanCenter recipes, which are typically not opened by developers to touch the source code)

SpaceIm replied:
There is no breaking change, I've not changed toolchain files.

Here is another proposal:

- add something like `props_autoinjection` argument to `build()` method of `MSBuild`, `False` by default.
- add properties dict attribute to `MSBuild`
- when `props_autoinjection` is False, current behavior is kept, so no breaking change at all.
- when `props_autoinjection` is True:
  - it collects root props of `MSBuildToolchain` & `MSBuildDeps` and inject them to `/p:ForceImportBeforeCppTargets`.
  - it also injects `/p:<property>=<value>` to msbuild command line
- keep previous tests as is, an write other for the `props_autoinjection=True` flavor.

SpaceIm replied:
Regarding properties, to avoid duplication between toolchain and MSBuild, there is a more complex solution: gather all properties by parsing all props of a specific configuration programmatically, and inject them in command line if props_autoinjection is True.

memsharded replied:
Yes, sorry, I meant the ``/p:ForceImportBeforeCppTargets`` is a breaking change when applied to ``conandeps.props``. For the toolchain, the main issue is that it can easily be hiding errors for developers opening the IDE

Yes, I think this approach would be better. Just a couple of suggestions:
- Make explicit the intention, ``props_autoinjection`` is a bit generic, something like ``force_imports`` or ``force_import_generated_files``
- Generic property injection is a different feature and ``MSBuildToolchain`` already provides injection of properties via toolchain (so they work for developers opening the IDE, without needing the Conan ``build()`` injection), so this should be a separate discussion if there is a use case for it.

SpaceIm replied:
> Generic property injection is a different feature and MSBuildToolchain already provides injection of properties via toolchain (so they work for developers opening the IDE, without needing the Conan build() injection), so this should be a separate discussion if there is a use case for it.

As I said in https://github.com/conan-io/conan/pull/12817#discussion_r1059935272, properties defined in toolchain files are not honored when injected through `/p:ForceImportBeforeCppTargets`, so for me it's related. `/p:ForceImportBeforeCppTargets` would be an internal detail of `force_import_generated_files =True`, the goal of this argument is to honor everything defined in these props, by any means.

memsharded commented on conans/test/functional/toolchains/microsoft/test_msbuilddeps.py:
```diff
@@ -115,9 +115,6 @@
   </ImportGroup>
   <ImportGroup Label="Shared">
   </ImportGroup>
-  <ImportGroup Label="PropertySheets">
```
This test is testing exactly that different subprojects can depend on individual props, without needing to link all dependencies, please leave this test not modified.

memsharded reviewed (changes requested):
There are different things proposed in this PR, and they probably require separate discussions:
- Using /p:ForceImportBeforeCppTargets to automatically inject property files: At least requires an opt-out, but probably an opt-in is better, we need to discuss this.
- Passing the PlatformToolset in the command line is hiding if the PlatformToolset definition from the toolchain works.
- Making visible for users the ``vs_toolset()`` function

It seems that the best would be to make the ``p:ForceImportBeforeCppTargets`` available as opt-in, and cover it with new test, instead of modifying existing tests.

SpaceIm committed:
allow to disable automatic injection of dependencies props

SpaceIm committed:
Revert "allow to disable automatic injection of dependencies props"

This reverts commit 86762f7c2c2b97e4e82159075ca24a086d511077.

SpaceIm committed:
Revert "move /p options first"

This reverts commit d0fc429198fa8a95f18d937b8db15ad82eb3ad97.

SpaceIm committed:
Revert "force PlatformToolset from msbuild command line"

This reverts commit f6da5206b86b8ab182f82a99293b2e3e48c5e4ae.

SpaceIm committed:
Revert "adapt template & tests"

This reverts commit 297677677df8ef440251056a7603003c181e09e6.

SpaceIm committed:
add force_import_generated_files argument to MSBuild.build()

this argument is disabled by default. When enabled, it automatically injects props files generated by MSBuildToolchain and MSBuildDeps.
A new attribute configuration is added to MSBuild, because we have to extract properties from MSBuildToolchain for the configuration/platform.

SpaceIm committed:
typo

memsharded commented on conan/tools/microsoft/visual.py:
```diff
@@ -268,3 +268,14 @@ def is_msvc_static_runtime(conanfile):
+
+
+def msbuild_arch(conanfile):
+    if conanfile.settings.get_safe("os") == "WindowsCE":
+        return conanfile.settings.get_safe("os.platform")
```
Those 2 lines are new logic not previously in the code, and those are untested. Better leave them as-is, and if necessary, treat the problem of WindowsCE in a different PR.

SpaceIm replied:
reverted

SpaceIm committed:
take into account namespace in props files

memsharded commented on conan/tools/microsoft/msbuild.py:
```diff
@@ -48,11 +47,75 @@ def command(self, sln, targets=None):
         return NotImplementedError("get_version() method is not supported in MSBuild "
                                    "toolchain helper")
+
+    def _get_concrete_props_file(self, root_props_file):
```
I am not sure why the specific file needs to be passed, if the general toolchain file already contains conditionals to pick the right configuration one.

SpaceIm replied:
> if the general toolchain file already contains conditionals to pick the right configuration one.

Because it doesn't work.
I'm testing with several CCI recipes for the moment, and properties are ignored when they come only from props file (msbuild complains about PlatformToolset since it's not overridden, and when there is a hardcoded WholeProgramOptimization I can't disable it through a custom property in MSBuildToolchain). When I extract these values from toolchain file and inject them in command line, it works.

SpaceIm replied:
(obviously, I would be more than happy if a smart solution not involving to parse xml files could be found)

memsharded replied:
I still see a lot of complexity in this functionality.
Wouldn't it be way more straightforward to just inject the ``<import>`` section in the ``vcxproj`` directly?

memsharded commented on conan/tools/microsoft/msbuild.py:
```diff
@@ -48,11 +47,75 @@ def command(self, sln, targets=None):
+        properties = {}
+
+        # Get properties from props file of configuration and platform
+        concrete_props_file = self._get_concrete_props_file(root_props_file)
```
Maybe this can be avoided with ``ForceImportAfterCppTargets``?

SpaceIm replied:
Yes maybe, I can try

SpaceIm replied:
Nop, it doesn't help :(

memsharded reviewed (commented):
It seems it is getting better, but I am still a bit concerned about ``MSBuild`` doing too much, parsing XML files to extract the properties...

Also, this new functionality should be covered by new tests, please add them when this converges to something that we agree on. Many thanks!

SpaceIm committed:
revert WindowsCE logic

SpaceIm committed:
restore difference between msbuild_arch injected by MSBuild and the one used for conditions in props file

SpaceIm commented on conan/tools/microsoft/msbuild.py:
```diff
@@ -16,10 +22,18 @@ def msbuild_arch(arch):
 
 
+def msbuild_arch_to_conf_arch(arch):
+    return {'x86': 'Win32',
```
I'm a little bit surprised of this difference. In conan-center, recipes using legacy MSBuild were always overriding x86 to pass Win32. Is it not a mistake?

SpaceIm committed:
typo

SpaceIm committed:
handle platform overridden as Win32 in MSBuild

SpaceIm committed:
improve robustness

SpaceIm committed:
test force_import_generated_files=True flavor

SpaceIm committed:
fix tests

SpaceIm committed:
fix again

SpaceIm commented:
```
E                   TypeError: test_toolchain_win_vs2017() missing 5 required positional arguments: 'compiler', 'version', 'runtime', 'cppstd', and 'force_import_generated_files'
```

I don't understand. What am I missing?

EDIT: ok there is a mixup of Unittest, parameterized and pytest.

SpaceIm committed:
no mixup of unittest, parameterized and pytest

SpaceIm committed:
more fix

SpaceIm committed:
fix again

SpaceIm committed:
fix test again

SpaceIm committed:
minor change

SpaceIm commented:
@memsharded I've updated a functional test_msbuild test to test both flavors of force_import_generated_files.

SpaceIm committed:
simplify test_msbuild

SpaceIm committed:
fix side effect in intel msbuild test

SpaceIm committed:
in test_msbuild, change PlatformToolset in project file to an old one to really test that it can be overridden

SpaceIm committed:
cleanup

SpaceIm committed:
factorize default platform computation in MSBuildToolchain & MSBuildDeps

SpaceIm commented on conan/tools/microsoft/toolchain.py:
```diff
@@ -52,27 +61,23 @@ def __init__(self, conanfile):
         self.cflags = []
         self.ldflags = []
         self.configuration = conanfile.settings.build_type
+        self.platform = arch_to_vcproj_platform(str(conanfile.settings.arch))
```
I've exposed platform attribute in MSBuildToolchain because it was exposed in MSBuildDeps and they share the same underlying logic, but I'm not sure it makes sense. Is it really customizable in MSBuild files?

memsharded replied:
I am ok with it, I think it is customizable, but in the worst case, it is not critical that it is exposed, just don't change it if you don't need to change it.

memsharded commented on conans/test/functional/toolchains/microsoft/test_msbuild.py:
```diff
@@ -1,13 +1,10 @@
-from parameterized import parameterized, parameterized_class
 
-from conan.tools.microsoft.visual import vcvars_command
+from conan.tools.microsoft.visual import msvc_version_to_vs_ide_version, vcvars_command
```
Please do not modify existing tests. There is a risk of breaking some functionality. The feature that you are proposing is a opt-in, so the previous tests shouldn't be modified, and adding a new test that uses the new functionality is the way to go.

SpaceIm replied:
It's deeply modified because this test was not easily extendable, but it tests the same things (and even more, properties injection was not tested actually, and it allows to test other Visual Studio versions than 14, it's an issue for local tests). It seems weird to add a separate test with a risk to have discrepancies between both flavors.
But yeah if you want I can move this to another file, it's just a lot of duplicated code (specially this big vcxproj file).

memsharded commented on conan/tools/microsoft/msbuild.py:
```diff
@@ -48,11 +68,83 @@ def command(self, sln, targets=None):
+
+        # MSBuildToolchan must be in generators for this MSBuild mode
+        msbuildtoolchain_file = os.path.join(self._conanfile.generators_folder, MSBuildToolchain.filename)
+        if not os.path.exists(msbuildtoolchain_file):
```
It is possible that some users want to inject only dependencies information, but they don't want to use the toolchain at all? Sounds very possible.

SpaceIm replied:
Il would make little sense. It would be like allowing `CMake` build helper to not consume `conan_toolchain.cmake`.

memsharded commented on conan/tools/microsoft/msbuild.py:
```diff
@@ -16,10 +22,21 @@ def msbuild_arch(arch):
             'armv8': 'ARM64'}.get(str(arch))
 
 
+def msbuild_arch_to_conf_arch(arch):
```
If used only once, then it is better to keep it local initially, until repetition pattern clearly arises (ideally at least 3 occurrences)

SpaceIm replied:
It's not exposed in `__init__.py` of `microsoft` folder, so it's not public according to conan documentation. What should I do? There are other non-public functions in this file actually.

memsharded reviewed (commented):
I am still seeing quite a complexity in this PR that I feel it is not great for future issues and maintenance.
For all company usage, adding the .props files to the projects is necessary anyway.
For ConanCenter usage, I think there are way simpler alternatives:
- Patching the ``vcxproj`` to inject the .props files, in the recipe
- Implementing the injection of .props files in the ``vcxproj`` with a helper.
In all cases, the functionality should be oriented to get to the desired recommended state: a ``vcxproj`` that imports the ``.props`` files in the right location. Having to parse the xml of the projects to extract import files, to extract properties, to pass them in command line is too complex IMO.

Asking for further reviews.

SpaceIm commented:
>I am still seeing quite a complexity in this PR that I feel it is not great for future issues and maintenance.
> For all company usage, adding the .props files to the projects is necessary anyway.

It's necessary for company usage when they work in Visual Studio IDE, but it can be an issue if they want to package an external dependency with build files not under their control, like conan-center.

> For ConanCenter usage, I think there are way simpler alternatives:
> 
> Patching the vcxproj to inject the .props files, in the recipe

It's what I've done in several recipes to fix them (and it was not obvious to find where this file should be injected in vcxproj files, it's very important actually and not documented in conan doc). And before I did, all packages were broken (even the template written by conan-center maintainers) due to the confusion introduced by this discrepancy behavior between MSBuildToolchain/MSBuild and other build helpers families. It's a good indication that it's not intuitive and there is something wrong.

santhonisz commented:
Just wondering if an alternate/more viable solution to this could be something along the lines of https://github.com/conan-io/conan/issues/3859#issuecomment-1280544191? 

This would be also be a similar approach to what is suggested for `vcpkg`: https://vcpkg.io/en/docs/users/buildsystems/msbuild-integration.html#import-props-and-targets

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=12817) <br/>All committers have signed the CLA.

memsharded commented:
Following up on some older PRs.
I am going to close this as not planned, it has too much complexity, doing too many different things in the same PR, modifying tests that shouldn't be modified, so the risk is high. Thanks very much for your contribution anyway, it is still interesting and can be useful for future improvements.

If we want to pursue improvement in this direction, it should be done in smaller and more focused steps.
A first possible small step would be doing just the ``cmd_args.append(f"/p:ForceImportBeforeCppTargets=\"{';'.join(props_paths)}\"")`` argument appending, with some minimal new test and not modifying existing ones. If some things do not work because the project hardcodes toolsets or something, it is not a problem, that could be a different thing, better handle it separately.
